### PR TITLE
Support nested generics

### LINKED ISSUE #362: Nested generics
Trying to call a generic method inside a generic type leads to a failure. It seems like we are only using the outer-most layer of generic arguments for specialization/substitution.

### PR DESCRIPTION
**Pull request by csyonghe**
fixes #362 

This pull request contains various bug fixes to get nested generics to work.

### CODE REVIEW

File: source/slang/ir.cpp [hunk 1]
```diff
@@ -4601,7 +4579,8 @@ namespace Slang
         // At this point we've created a new non-generic function,
         // which means we should add it to our work list for
         // subsequent processing.
-        sharedContext->workList.Add(specFunc);
+        if (specFunc->specializedGenericLevel == -1)
+            sharedContext->workList.Add(specFunc);
```
Comment (tfoleyNV) [inline source/slang/ir.cpp]: This is extremely minor, but I would have assumed that a `grnericSpecializationLevel` of zero would represent the non-generic case, not `-1`.

File: source/slang/ir.cpp [hunk 2]
```diff
@@ -4760,14 +4739,16 @@ namespace Slang
...
                                 auto genericFunc = (IRFunc*)genericVal;
-                                if (!genericFunc->genericDecl)
+                                if (!genericFunc->getGenericDecl())
                                     continue;
 
                                 // Okay, we have a candidate for specialization here.
                                 //
                                 // We will first find or construct a specialized version
                                 // of the callee funciton/
+                                auto originalFunc = dumpIRFunc(genericFunc);
```
Comment (tfoleyNV) [inline source/slang/ir.cpp]: I assume this dumping logic shouldn’t be included, and it was just for debugging?

File: source/slang/syntax.cpp [hunk 3]
```diff
@@ -1966,5 +1949,39 @@ void Type::accept(IValVisitor* visitor, void* extra)
             rs = combineHash(rs, globalGenParamSubstitutions->GetHashCode());
         return rs;
     }
+
+    RefPtr<Substitutions> cloneSubstitutionNonRecursive(RefPtr<Substitutions> subst)
```
Comment (tfoleyNV) [inline source/slang/syntax.cpp]: Where does this get used?

File: source/slang/ir.cpp (hunk 1 above)
Comment (csyonghe) [reply source/slang/ir.cpp]: Yeah, but `genericSpecializationLevel` is used as an index into the `genericDecls` list, where the last element is the outer most decl and first element is inner most generic decl. So a non-generic function will have an empty `genericDecls` list and `genericSpecializationLevel==-1`.

**Pull request merged**